src/main/scala/cognite/spark/NumericDataPointsRdd.scala

```diff
@@ -19,32 +42,257 @@ case class NumericDataPointsRdd(
+    Granularity(8, ChronoUnit.DAYS),
+    Granularity(1, ChronoUnit.DAYS),
+    Granularity(12, ChronoUnit.HOURS),
+    Granularity(6, ChronoUnit.HOURS),
+    Granularity(3, ChronoUnit.HOURS),
+    Granularity(1, ChronoUnit.HOURS),
+    Granularity(30, ChronoUnit.MINUTES),
+    Granularity(1, ChronoUnit.MINUTES),
+    Granularity(30, ChronoUnit.SECONDS),
+    Granularity(1, ChronoUnit.SECONDS)
+  )
+
+  private def smallEnoughRanges(
+      id: Long,
+      start: Instant,
+      end: Instant,
+      granularities: Seq[Granularity] = granularitiesToTry): IO[Seq[Range]] =
+    granularities match {
+      case granularity +: moreGranular =>
+        client.dataPoints
+          .queryAggregatesById(
+            id,
+            start,
+            end,
+            granularity.toString,
+            Seq("count"),
+            limit = Some(10000)
+          )
+          .flatMap { aggregates =>
+            val counts = aggregates("count")
+            if (counts.map(_.value).max > partitionSize && moreGranular.nonEmpty) {
+              smallEnoughRanges(id, start, end, moreGranular)
+            } else {
+              IO.pure(countsToRanges(id, counts, granularity))
+            }
+          }
+    }
+
+  private def getFirstAndLastConcurrently(ids: Vector[Long], start: Instant, end: Instant) = {
+    val firsts = ids.map { id =>
+      client.dataPoints
+        .queryById(
+          id,
+          start,
+          end,
+          limit = Some(1)
+        )
+        .map(_.headOption)
+        .map(p => id -> p)
+    }.parSequence
+    val lasts = client.dataPoints.getLatestDataPointsByIds(ids)
+    (firsts, lasts).parMapN {
+      case (f, l) =>
+        f.map {
+          case (id, first) =>
+            (id, first, l.getOrElse(id, None))
+        }
+    }
+  }
+
+  private def rangesToBuckets(ranges: Seq[Range]) = {
+    // Fold into a sequence of buckets, where each bucket has some ranges with a
+    // total of data points <= bucketSize
+    val bucketsFold = ranges.foldLeft((Seq.empty[Bucket], Seq.empty[Range], 0L)) {
+      case ((buckets, ranges, sum), r) =>
+        val sumTotal = sum + r.count
+        if (sumTotal > bucketSize) {
+          // Create a new bucket from the ranges, and assign an index of "1" temporarily.
+          // We'll change the index before returning the buckets.
+          (Bucket(1, ranges) +: buckets, Seq(r), r.count)
+        } else {
+          (buckets, r +: ranges, sumTotal)
+        }
+    }
+
+    // If there are any non-empty ranges left, put them in a final bucket.
+    val buckets = bucketsFold match {
+      case (buckets, _, 0) => buckets
+      case (buckets, ranges, _) => Bucket(1, ranges) +: buckets
+    }
+
+    buckets.zipWithIndex.map {
+      case (bucket, index) => bucket.copy(index = index)
+    }
+  }
+
+  private def buckets(
+      ids: Seq[Long],
+      externalIds: Seq[String],
+      start: Instant,
+      end: Instant): Seq[Bucket] = {
+    val firstLatest = Stream
```

Comment:
It's a bit of `parEvalMapUnordered` but also because we want to start processing each chunk asap as it's finished. Without `fs2` we don't get that, we'd have to resort to non-lazy `Vector` or similar as far as I can tell (I tried).